react/features/whiteboard/middleware.ts

```diff
@@ -0,0 +1,166 @@
+// @ts-ignore
+import { MiddlewareRegistry, StateListenerRegistry } from '../base/redux';
+import { HIDE_WHITEBOARD, OPEN_WHITEBOARD, SHOW_WHITEBOARD } from './actionTypes';
+import { isWhiteboardOpen, isWhiteboardPresent } from './functions';
+import { WHITEBOARD, WHITEBOARD_ID, WHITEBOARD_PARTICIPANT_NAME } from './constants';
+import { resetWhiteboard, setupWhiteboard } from './actions';
+import { getCurrentRoomId } from '../breakout-rooms/functions';
+
+// @ts-ignore
+import { addStageParticipant } from '../filmstrip/actions.web';
+
+// @ts-ignore
+import { isStageFilmstripAvailable } from '../filmstrip/functions';
+
+const focusWhiteboard = (store: IStore) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const conference = getCurrentConference(state);
+    const stageFilmstrip = isStageFilmstripAvailable(state);
+    const isPresent = isWhiteboardPresent(state);
+
+    if (!isPresent) {
+        dispatch(participantJoined({
+            conference,
+            id: WHITEBOARD_ID,
+            isFakeParticipant: true,
+            isWhiteboard: true,
+            name: WHITEBOARD_PARTICIPANT_NAME
+        }));
+    }
+    if (stageFilmstrip) {
+        dispatch(addStageParticipant(WHITEBOARD_ID, true));
+    } else {
+        dispatch(pinParticipant(WHITEBOARD_ID));
+    }
+};
+
+/**
+ * Handles the whiteboard state.
+ *
+ * @param {Store} store - The redux store.
+ * @param {Object} attributes - The command attributes.
+ * @returns {void}
+ */
+const handleWhiteboardStatus = (store: IStore, { roomId, roomKey }: { roomId: string; roomKey: string; }) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const isOpen = isWhiteboardOpen(state);
+
+    if (isOpen) {
+        return;
+    }
+
+    focusWhiteboard(store);
+    dispatch(setupWhiteboard({
+        collabDetails: {
+            roomId,
+            roomKey
+        }
+    }));
+};
+
+/**
+ * Sends the WHITEBOARD command.
+ *
+ * @returns {void}
+ */
+const sendWhiteboardCommand = ({ conference, roomId, roomKey }: {
+    conference: any;
+    roomId?: string;
+    roomKey?: string;
+}): void => {
+    conference.sendCommand(WHITEBOARD, {
+        attributes: {
+            roomId,
+            roomKey
+        }
+    });
+};
+
+/**
+ * Middleware which intercepts whiteboard actions to handle changes to the related state.
+ *
+ * @param {Store} store - The redux store.
+ * @returns {Function}
+ */
+MiddlewareRegistry.register((store: IStore) => (next: Function) => async (action: any) => {
+    const { dispatch, getState } = store;
+    const state = getState();
+    const localParticipant = getLocalParticipant(state);
+
+    switch (action.type) {
+    case OPEN_WHITEBOARD: {
+        if (!isParticipantModerator(localParticipant) && !isEveryoneModerator(state)) {
+            return;
+        }
+
+        const conference = getCurrentConference(state);
+        const collabDetails = await generateCollaborationLinkData();
```

Comment:
What is async about this function? Aren't they 2 random strings?

Reply:
This is an async method exported from excalidraw that eventually uses https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/generateKey